WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewModel.kt

```diff
@@ -0,0 +1,252 @@
+import org.wordpress.android.ui.sitecreation.verticals.NewSiteCreationVerticalsViewModel.VerticalsListItemUiState.VerticalsModelUiState
+import org.wordpress.android.viewmodel.SingleLiveEvent
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.coroutines.experimental.CoroutineContext
+
+private const val throttleDelay: Int = 500
+
+class NewSiteCreationVerticalsViewModel @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val fetchVerticalsHeaderInfoUseCase: FetchVerticalsHeaderInfoUseCase,
+    private val fetchVerticalsUseCase: FetchVerticalsUseCase,
+    @Named(IO_DISPATCHER) private val IO: CoroutineContext,
+    @Named(MAIN_DISPATCHER) private val MAIN: CoroutineContext
+) : ViewModel(), CoroutineScope {
+    private val job = Job()
+    private var fetchVerticalsJob: Job? = null
+    override val coroutineContext: CoroutineContext
+        get() = IO + job
+    private var isStarted = false
+
+    private val _uiState: MutableLiveData<VerticalsUiState> = MutableLiveData()
+    val uiState: LiveData<VerticalsUiState> = _uiState
+
+    private var listState: ListState<VerticalModel> = ListState.Init()
+    private lateinit var headerInfo: DummyVerticalsHeaderInfoModel
+
+    private val _clearBtnClicked = SingleLiveEvent<Void>()
+    val clearBtnClicked = _clearBtnClicked
+
+    init {
+        dispatcher.register(fetchVerticalsUseCase)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        dispatcher.unregister(fetchVerticalsUseCase)
+    }
+
+    fun start() {
+        if (isStarted) {
+            return
+        }
+        isStarted = true
+        fetchHeaderInfo()
+    }
+
+    private fun fetchHeaderInfo() {
+        launch {
+            withContext(MAIN) {
+                updateUiStateToFullScreenProgress()
+            }
+            val headerInfoEvent = fetchVerticalsHeaderInfoUseCase.fetchVerticalHeaderInfo()
+            withContext(MAIN) {
+                onHeaderInfoFetched(headerInfoEvent)
+            }
+        }
+    }
+
+    private fun onHeaderInfoFetched(event: DummyOnVerticalsHeaderInfoFetched) {
+        if (event.isError) {
+            updateUiStateToFullScreenError()
+        } else {
+            headerInfo = event.headerInfo!!
+            updateUiStateToContent("", ListState.Ready(emptyList()))
+        }
+    }
+
+    fun onFetchHeaderInfoRetry() {
+        fetchHeaderInfo()
+    }
+
+    fun onClearTextBtnClicked() {
+        _clearBtnClicked.call()
+    }
+
+    fun updateQuery(query: String, delay: Int = throttleDelay) {
+        fetchVerticalsJob?.cancel() // cancel any previous requests
+        if (query.isNotEmpty()) {
+            fetchVerticals(query, delay)
+        } else {
+            updateUiStateToContent(query, ListState.Ready(emptyList()))
+        }
+    }
+
+    private fun fetchVerticals(query: String, throttleDelay: Int) {
+        fetchVerticalsJob = launch {
```

Comment:
Don't we need to check for `isActive` in this `coroutine` since it's cancellable? I thought it continues with the job otherwise, but I might be wrong about that.

This is one of the reasons why I am kind of surprised about how we are handling the cancellation, because I'd imagine that the only part we actually want to cancel & restart is fetching the verticals and the rest should continue as is. We still update the UI to `Loading`, we still fetch verticals and we still update it to `Success` or `Error` when the fetch actually completes. We shouldn't really need to care about how many times the fetch action was dispatched or how long it took, except for when the query is fully deleted in which case we can set it to `Ready` again.

Just thinking out loud a bit as I try to wrap my head around this PR, please feel free to ignore the second part of this comment if it doesn't make sense.

Reply:
>Don't we need to check for isActive in this coroutine since it's cancellable? I thought it continues with the job otherwise, but I might be wrong about that.

I'm not sure I understand the question correctly. The cancel is called in `updateQuery(..)` and the coroutine is automatically cancelled when it's in `delay` or in `suspendCancellableCoroutine()` inside the UseCase. If it was actively computing all the time we'd really have to check for `isActive`.

Reply:
>This is one of the reasons why I am kind of surprised about how we are handling the cancellation, because I'd imagine that the only part we actually want to cancel & restart is fetching the verticals and the rest should continue as is. We still update the UI to Loading, we still fetch verticals and we still update it to Success or Error when the fetch actually completes. 

I agree, but it's how it basically works except we send the Loading state for each request, which is unnecessary. Do you have any proposal how to improve it?

>We shouldn't really need to care about how many times the fetch action was dispatched or how long it took, except for when the query is fully deleted in which case we can set it to Ready again.

I'm not sure I follow. In what way we care how many times and how long the fetch action was dispatched?

Reply:
> I agree, but it's how it basically works except we send the Loading state for each request, which is unnecessary. Do you have any proposal how to improve it?

I'll try out what I had in mind and add a diff for it.

> I'm not sure I follow. In what way we care how many times the fetch action was dispatched or how long it took?

We don't care about how long it takes, but we do update the views each time the job is launched.

Reply:
I'm sorry, but I'm probably missing something :(. 
- This approach will still be updating the Views each time the the job is launched, right?
- `TODO: Use a cancelable job and add a delay. ` I'm still new to coroutines. Are you proposing to create a nested coroutine in the UseCase and cancel just it's job? Tbh I'm not sure how canceling just a nested job works - what does the parent do when it's children are canceled?

As I said, I'm probably missing something. But I feel like we just moved responsibility for throttling from the VM to the UseCase. Which I'm not sure is the right direction.